Transition to R8 in full mode

MortimerGoro opened the pull request:
This PR adds the proguard rules required to make R8 work in full mode. It will allow to reduce compiling times significantly.

In addition, we were not doing correct minification before. We had a rule to keep all classes. I removed that so it has required to add some extra rules such as `WidgetPlacement` class which fields are uses from native.

One of the problematic dependencies was `mozillaspeechlibrary`. I added some rules here but also pushed a PR to add the rules upstream: https://github.com/mozilla/androidspeech/pull/17

Other problematic dependency was Gecko. The fenix tem has already enabled full mode: https://github.com/mozilla-mobile/fenix/blob/master/gradle.properties. But it triggered a weird error in FxR. After some tries I fixed it keeping everything inside `public class org.mozilla.gecko.**`. I'll follow up with the GV team and file a bug. But for now we can use this rule (keep in mind we were keeping everything before...)

MortimerGoro commented:
Fixes #1190

MortimerGoro commented:
I found a better way to fix the Gecko rules. Posted a Gecko patch here: https://bugzilla.mozilla.org/show_bug.cgi?id=1567232

We can remove the extra gecko rules here when the GV patch lands

bluemarvin commented:
> I found a better way to fix the Gecko rules. Posted a Gecko patch here: https://bugzilla.mozilla.org/show_bug.cgi?id=1567232
> 
> We can remove the extra gecko rules here when the GV patch lands

Looks like the GV patch has already landed, do we just want to wait for the GV version bump to land this?

MortimerGoro commented:
I removed the rules we added to GV consumer proguard file. Let's land the GV version bump https://github.com/MozillaReality/FirefoxReality/pull/1441 and then this

bluemarvin commented:
APK is about 3MB smaller it seems.

bluemarvin commented:
Oculus crash: https://crash-stats.mozilla.com/report/index/308c9afa-c805-42b9-a354-cef020190722

MortimerGoro commented:
I've fixed the Oculus crash in https://bugzilla.mozilla.org/show_bug.cgi?id=1567232

Also added new rules for HTC devices in FxR proguard file

bluemarvin commented:
I now get the same crash on HTC as Oculus so need to wait for GV to get updated.

MortimerGoro commented on app/proguard-rules.pro:
```diff
@@ -55,6 +73,4 @@
 -optimizations !code/simplification/arithmetic,!code/allocation/variable
 -keepattributes *
--keep class !org.mozilla.gecko.Speech* { *; }
--keepclassmembers class !org.mozilla.gecko.Speech*, ** { *; }
```
That rule kept all classes but that one. Now we have implemented better minification and only have rules to keep the needed classes
